pkg/mesh/td.go

```diff
@@ -0,0 +1,227 @@
+	if err != nil {
+		return "", fmt.Errorf("could not determine project number from metadata server: ", err)
+	}
+	return projectNumber, nil
+}
+
+func (td *TdSidecarEnv) fetchNodeID() (string, error) {
+	hostName, hostErr := os.Hostname()
+	if hostErr != nil {
+		return "", hostErr
+	}
+	ips, ipErr := net.LookupIP(hostName)
+	if ipErr != nil {
+		return "", ipErr
+	}
+	return fmt.Sprintf("%s~%s", uuid.New().String(), ips[0].String()), nil
+}
+
+// Prepare TdSidecarEnv by using Metadata server and environment variables.
+func (td *TdSidecarEnv) prepare() {
+	// We use the project number that this cloud run service is running under as a default.
+	if projectNumber, err := td.fetchProjectNumber(); err != nil {
+		log.Println("Unable to auto-generate project_number.")
+	} else {
+		td.ProjectNumber = projectNumber
+	}
+
+	td.loadFromEnv()
+
+	if nodeID, err := td.fetchNodeID(); err != nil {
+		td.NodeID = fmt.Sprintf("%s~%s", uuid.New().String(), "127.0.0.1")
+		log.Println("Unable to generate proper nodeID, using: ", td.NodeID)
+	} else {
+		td.NodeID = nodeID
+	}
+
+	if zone, err := td.fetchZone(); err != nil {
+		td.EnvoyZone = "cloud-run-cluster"
+		log.Println("Unable to generate proper zone info, using: ", td.EnvoyZone)
+	} else {
+		td.EnvoyZone = zone
+	}
+}
+
+func (td *TdSidecarEnv) getIPTablesInterceptionEnvVars() []string {
+	envs := []string{
+		"TRAFFIC_DIRECTOR_GCE_VM_DEPLOYMENT_OVERRIDE=true",
+		"DISABLE_REDIRECTION_ON_LOCAL_LOOPBACK=true",
+		fmt.Sprintf("%s=%s", "ENVOY_DNS_PORT", td.EnvoyDnsPort),
+	}
+	return envs
+}
+
+func (td *TdSidecarEnv) logLevel() string {
+	return td.LogLevel
+}
+
+func (td *TdSidecarEnv) envoyPort() string {
+	return td.EnvoyPort
+}
+
+func (td *TdSidecarEnv) envoyAdminPort() string {
+	return td.EnvoyAdminPort
+}
+
+func (td *TdSidecarEnv) projectNumber() string {
+	return td.ProjectNumber
+}
+
+func (td *TdSidecarEnv) networkName() string {
+	return td.NetworkName
+}
+
+func (td *TdSidecarEnv) serviceCIDR() string {
+	return td.ServiceCidr
+}
+
+func (td *TdSidecarEnv) envoyLogDirectory() string {
+	return td.LogDirectory
+}
+
+// Reads bootstrap template found in absolute path templatePath and replaces strings.
+func (td *TdSidecarEnv) prepareTrafficDirectorBootstrap(templatePath string, outputPath string) error {
+	data, err := os.ReadFile(templatePath)
+	if err != nil {
+		return err
+	}
+
+	template := string(data)
+	template = strings.ReplaceAll(template, "ENVOY_NODE_ID", td.NodeID)
+	template = strings.ReplaceAll(template, "ENVOY_ZONE", td.EnvoyZone)
+	template = strings.ReplaceAll(template, "VPC_NETWORK_NAME", td.NetworkName)
+	template = strings.ReplaceAll(template, "CONFIG_PROJECT_NUMBER", td.ProjectNumber)
+	template = strings.ReplaceAll(template, "ENVOY_PORT", td.EnvoyPort)
+	template = strings.ReplaceAll(template, "ENVOY_ADMIN_PORT", td.EnvoyAdminPort)
+	template = strings.ReplaceAll(template, "XDS_SERVER_CERT", td.XdsServerCert)
```

Comment:
How would the XDS_SERVER_CERT be located? You can't mount.

Reply:
TD hard codes XDS_SERVER_CERT to '/etc/ssl/certs/ca-certificates.crt' and expects it to exist in the image. I think for gcr.io/istio-testing/proxyv2:latest, this file exists. stat /etc/ssl/certs/ca-certificates.crt
 ---> Running in 6ce8cc5d2e37
  File: /etc/ssl/certs/ca-certificates.crt
  Size: 199113    	Blocks: 392        IO Block: 4096   regular file
Device: 2fh/47d	Inode: 2885566     Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-10-06 19:03:24.000000000 +0000
Modify: 2021-10-06 19:03:24.000000000 +0000
Change: 2021-11-10 15:58:21.870307836 +0000